### ISSUE TITLE
Allow daemons to exit immediately when the connection terminates?

### ISSUE THREAD

**Issue by wlandau** [shikokuchuo/mirai#87]
If a daemon is running a long task and `daemons(n = 0L)` is called, the daemon continues to run the task and does not exit until the task is finished. A reprex is below.

```r
library(mirai)
library(ps)
url <- "ws://127.0.0.1:5700"
daemons(n = 1L, url = url)
#> [1] 1
tasks <- replicate(2, mirai(Sys.sleep(100)))
launch_local(url = url)
Sys.sleep(2)
daemons(n = 0L)
#> [1] 0
# Check the daemon PID on e.g. htop.
```

As you explained in #86, this behavior is desirable in many situations: e.g. if a task has intermediate checkpoints and needs a chance to log them. In `targets` pipelines, however, the extra execution time is not useful because the metadata cannot be logged if the central R process is interrupted. In addition, I worry about mounting costs on expensive cloud services. So I think it would be useful to be able to configure the daemons to exit as soon as the TCP connections are broken. c.f. https://github.com/wlandau/crew/issues/141

**Comment by shikokuchuo** [comment]
It might be possible to allow this through an argument at `daemon()`. That way `crew` or `targets` can set it if desired. I think I'll be constrained by what is allowed in package space though, meaning probably raising SIGINT to signal an interrupt. This will cause R evaluation to stop, but uninterruptible C code may still continue until a user interrupt check is performed. Hence, your other monitoring utilities will still likely be useful in case they need to perform 'hard' shutdowns.

**Comment by wlandau** [comment]
> It might be possible to allow this through an argument at daemon(). That way crew or targets can set it if desired. 

That would be such a huge help as far as safety is concerned!

>  I think I'll be constrained by what is allowed in package space though, meaning probably raising SIGINT to signal an interrupt.

Are you referring to the clause from https://cran.r-project.org/doc/manuals/R-exts.html#Writing-portable-packages?

> Under no circumstances should your compiled code ever call abort or exit: these terminate the user’s R process, quite possibly losing all unsaved work.

I think `daemon()` is a legitimate exception. In this case, losing unsaved work is the goal. For an automated pipeline, if the dispatcher exits, then all unsaved work from the daemon is moot at best, harmful at worst.

Also, when they say "the user’s R process", I think they envision a single interactive local R session with a backlog of unsaved work from multiple projects. By contrast, the daemon is external and limited in scope.

> This will cause R evaluation to stop, but uninterruptible C code may still continue until a user interrupt check is performed.

Yes, I suppose not every block of C code can check [`R_CheckUserInterrupt()`]( https://cran.r-project.org/doc/manuals/R-exts.html#Allowing-interrupts).

**Comment by shikokuchuo** [comment]
Thanks for the link, that’s really quite interesting! It led me to do more thinking:

First, I think you are familiar with how pipe events work, but just to re-cap, we are not actually sending a signal to tell the daemons to exit (this would not be universally reliable for any type of remote connection), rather we are setting it up to raise the signal on itself when the connection is broken.

With this in mind, `callr` isn’t really equivalent as it is narrower in scope – it only launches background R processes locally. It would probably be safe in this case for the child processes to be killed.

With the open architecture design of `mirai` however, I can’t expect `daemon()` to always be in such a ‘disposable’ background process. Even `crew` calls `daemon()` through its own function, and it is reasonable to expect that similar setups may have an interrupt handler (in the simplest form a `tryCatch(…, interrupt = )` for example) perhaps for logging, perhaps for other reasons (re-spawning being an obvious one).

At the end of the day, we are talking about the length of an individual `.Call()` that signals may not reach (if it just so happens to be within that call), and it has always been the recommended practice for long-running C functions to call `R_CheckUserInterrupt()` at intervals in any case.

So for something like `targets` I think it’s going to be more than enough – at least for the intended purpose. Thinking through the consequences, if it were the case that such a non-interrupted call caused some data to be wrongly overwritten, then forcefully killing that process mid-write could result in more serious harm including data corruption. SIGINT is much safer in this regard in at least allowing a graceful exit.

**Comment by wlandau** [comment]
> Thinking through the consequences, if it were the case that such a non-interrupted call caused some data to be wrongly overwritten, then forcefully killing that process mid-write could result in more serious harm including data corruption. SIGINT is much safer in this regard in at least allowing a graceful exit.

I see what you mean. SIGINT allows cleanup to happen and ensures that output storage is not left in a corrupted or unusable state. In that sense, SIGINT is a more graceful and robust solution.

On the other hand, I don't think an abrupt SIGKILL is really such a disaster. If a data file is corrupted, all one needs to do is rerun the original code that produces it. (And `targets` will probably skip everything else, so fixing the data will consume almost no resources.) 

On the other hand, runaway processes are truly catastrophic. On the cloud, they could easily burn through tens of thousands of dollars and tank the cloud budgets of departments like mine. The probability of disaster is not small: poorly-written C/C++ code is prevalent in statistical modeling packages, and most of my colleagues who use `targets` are not skilled in computing. And if it does happen, the cleanup is long and arduous.

I work with hundreds of clinical statisticians, and almost all of them have little experience in R, no experience in high-performance computing, and no familiarity with the cloud. But most still use `targets` to run large, complex, custom-coded simulations to design clinical trials. For now, we use an on-prem SGE cluster, and each targets pipeline uses anywhere from 100 to 200 persistent workers concurrently. It is trivially easy to view cluster jobs with `qstat` and terminate them all with a single call to `qdel`, but almost none of my colleagues know how, let alone run commands in the Linux terminal. 

On top of that, we are trying to move to a cloud platform with workers on AWS Batch. When I implement [job management utilities in `crew.aws.batch`](https://github.com/wlandau/crew.aws.batch#job-management) on Friday, I learned that there is no wholesale way to terminate a group of Batch jobs. To manually terminate a pipeline of 100-200 workers, one would need to go through each one individually and send a separate TerminateJob HTTP API request. Even for someone who knows what they are doing, it could easily take the better part of an hour to put out a fire.

I saw your work at https://github.com/shikokuchuo/nanonext/pull/25, and I am super grateful. SIGINT does get us most of the way there. Given how different our use cases are, I wonder if it would be possible to let the user supply the signal type and make SIGINT the default.

**Comment by shikokuchuo** [comment]
This capability has been implemented in `nanonext`, and will be exposed by allowing a signal to be supplied to the 'autoexit' argument of `daemon()`.

**Comment by wlandau** [comment]
@shikokuchuo, I can't thank you enough! Using only the tools available on the host platform, it is almost impossible to automatically control runaway processes in the general case. Every bit helps a ton.

I tested this functionality and built it into development `crew`. Everything works super well.

**Comment by shikokuchuo** [comment]
You're welcome!

**Comment by wlandau** [comment]
An update: as I was testing just now, I found out that `tools::SIGKILL` is actually `NA` on Windows. That led me to do some more digging on signals, and I learned `SIGKILL` on a parent process can lead to zombie child processes. Zombies would defeat the whole purpose of my original preference for `SIGKILL`, so I am moving away from it. Instead, `crew` now prefers SIGTERM because the request to kill the process is more direct. If for some reason that signal is not defined on the user's platform, then it uses SIGQUIT. And if SIGQUIT is not defined, it uses SIGINT as a last resort. I don't think I will expose the choice of signal to `crew` users.
